vnext/Desktop.IntegrationTests/WebSocketServer.cpp

```diff
@@ -0,0 +1,390 @@
+  m_buffer.consume(m_buffer.size());
+
+  m_stream->text(m_stream->got_text());
+  m_stream->async_write(buffer(m_message), bind_executor(*m_strand, std::bind(
+    &BaseWebSocketSession<SocketLayer>::OnWrite,
+    this->SharedFromThis(),
+    _1, // ec
+    _2  // transferred
+  )));
+}
+
+template<typename SocketLayer>
+void BaseWebSocketSession<SocketLayer>::OnWrite(error_code ec, size_t /*transferred*/)
+{
+  if (ec)
+    return; //TODO: fail
+
+  // Clear outgoing message contents.
+  m_message.clear();
+
+  Read();
+}
+
+#pragma endregion // BaseWebSocketSession
+
+#pragma region WebSocketSession
+
+WebSocketSession::WebSocketSession(ip::tcp::socket socket, WebSocketServiceCallbacks& callbacks)
+  : BaseWebSocketSession(callbacks)
+{
+  m_stream = std::make_shared<websocket::stream<boost::asio::ip::tcp::socket>>(std::move(socket));
+  m_strand = std::make_shared<strand<io_context::executor_type>>(m_stream->get_executor());
+}
+
+WebSocketSession::~WebSocketSession() {}
+
+#pragma region BaseWebSocketSession
+
+std::shared_ptr<BaseWebSocketSession<boost::asio::ip::tcp::socket>> WebSocketSession::SharedFromThis() /*override*/
+{
+  return this->shared_from_this();
+}
+
+#pragma endregion // BaseWebSocketSession
+
+#pragma endregion // WebSocketSession
+
+#pragma region SecureWebSocketSession
+
+SecureWebSocketSession::SecureWebSocketSession(ip::tcp::socket socket, WebSocketServiceCallbacks& callbacks)
+  : BaseWebSocketSession(callbacks)
+{
+  // Initialize SSL context.
+  string const cert =
+    "-----BEGIN CERTIFICATE-----\n"
+    "MIIDhjCCAm6gAwIBAgIJAPh+egUebaStMA0GCSqGSIb3DQEBCwUAMFgxCzAJBgNV\n"
+    "BAYTAlVTMRMwEQYDVQQIDApXYXNoaW5ndG9uMRAwDgYDVQQHDAdSZWRtb25kMRIw\n"
+    "EAYDVQQKDAlNaWNyb3NvZnQxDjAMBgNVBAsMBVJlYWN0MB4XDTE5MDYwMTA4MDcx\n"
+    "M1oXDTI5MDUyOTA4MDcxM1owWDELMAkGA1UEBhMCVVMxEzARBgNVBAgMCldhc2hp\n"
+    "bmd0b24xEDAOBgNVBAcMB1JlZG1vbmQxEjAQBgNVBAoMCU1pY3Jvc29mdDEOMAwG\n"
+    "A1UECwwFUmVhY3QwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQCgHKC9\n"
+    "XC+1FjDg2Xdmbpf3ILiJQtGFiD3WFb+duNXThOA0LY6gytVBY6LitAzB7Jm7spvx\n"
+    "BbA46kw8Dsmv02hI0diVzFKCB5bTTs0N8bgAsem6qvDpo/mvp2TtDU2J8o4RhMQ3\n"
+    "BQvdZUGgtH4VR2W4vIHufNjVUvN9hTL2eOBz0EYElsMSogG8f97N+m/7L6JeyjPo\n"
+    "kFwXYTFMjv3ihJmev/cBNkxuchLUT7NAc7bMCtmv5lzsKMKe6g0lUxDSBYxXztqU\n"
+    "l3huo2g990VbvTWH/lhz3bgdnon/AUKWBmS2eRmK9hH/rGlm1NeMCjexMZrYC3m8\n"
+    "vvfIR25plGmNjyQJAgMBAAGjUzBRMB0GA1UdDgQWBBTNLE3Nl0s3O40wDEXf9t/7\n"
+    "r6Y1QjAfBgNVHSMEGDAWgBTNLE3Nl0s3O40wDEXf9t/7r6Y1QjAPBgNVHRMBAf8E\n"
+    "BTADAQH/MA0GCSqGSIb3DQEBCwUAA4IBAQAbRL+69uZLW3Q6JIQ9qg3CPjm6574o\n"
+    "cOiyJ9DX49i23AyYNsCYulvLApHgrmltMJHrC4U7EasQCTtwpAMyLJDLIDdujSSs\n"
+    "ynSe5PcNeElLmEkH4PxyAFsb/2oWI2PSJh0vseuugUpeKTHJv5MPkLUm7WMLHYj5\n"
+    "uOQzYDwJ+IuveVzX5TRXtkY8jF9ErL4iF8eYTyp0ANeY11vJOAbd2FcAy5baYjB3\n"
+    "JVczHy7eegwfOQJFM9mIZE7+Ac0SaknC0Jun9486cJ0mvbdrMSd+vgF85OrpWwYU\n"
+    "ISfux0NDVN1gjvSgdkEB+CWnV1rNKsVKlg4psDkpq33AJEwnx3qNxtKG\n"
+    "-----END CERTIFICATE-----\n";
```

Comment:
OOI, where is this certificate coming from? Is it part of an established trust chain? Is it ok to have a hard-coded cert?

Reply:
I didn't realize this is a server solely for testing. Please disregard the above comment.